File: cpp/rocketmq-client-cpp-tests/cpp4.x/test/client/message/MessageBodyContentTest.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <string>
#include <mutex>
#include <chrono>
#include <thread>
#include "gtest/gtest.h"
#include "spdlog/spdlog.h"
#include "rocketmq/MQMessage.h"
#include "rocketmq/DefaultMQProducer.h"
#include "rocketmq/DefaultMQPushConsumer.h"
#include "rocketmq/DefaultMQPullConsumer.h"
#include "frame/BaseOperate.h"
#include "listener/MsgListener.h"
#include "resource/Resource.h"
#include "factory/ConsumerFactory.h"
#include "factory/ProducerFactory.h"
#include "utils/VerifyUtils.h"

extern std::shared_ptr<spdlog::logger> multi_logger;
extern std::shared_ptr<Resource> resource;

// Send normal message, setting message body with space character, expect consume success
TEST(MessageBodyContentTest, testMessageBodyContentIsSpace)
{
    std::string topic = getTopic(MessageType::NORMAL, "testMessageBodyContentIsSpace", resource->getBrokerAddr(), resource->getNamesrv(), resource->getCluster());
    std::string group = getGroupId("testMessageBodyContentIsSpace");
    ASSERT_NO_THROW({
        std::shared_ptr<MsgListener> msglistener = std::make_shared<MsgListener>();
        auto pushConsumer = ConsumerFactory::getPushConsumer(topic, group, "*", msglistener);
        std::this_thread::sleep_for(std::chrono::seconds(5));

        auto producer = ProducerFactory::getProducer(group);

        std::string body = " ";
        rocketmq::MQMessage msg(topic, "*", body);
        rocketmq::SendResult sendResult = producer->send(msg);

        ASSERT_EQ(sendResult.getSendStatus(), rocketmq::SendStatus::SEND_OK);

        std::this_thread::sleep_for(std::chrono::seconds(5));

        // std::vector<rocketmq::MQMessageExt> msgs = VerifyUtils::fetchMessages(pullConsumer, topic);
        auto msgs = msglistener->getMessages();

        ASSERT_EQ(msgs.size(), 1);
        ASSERT_EQ(msgs[0].getBody(), body);

        pushConsumer->shutdown();
        producer->shutdown();
    });
}

// Send normal message, setting message body with chinese character, expect consume success
TEST(MessageBodyContentTest, testMessageBodyContentIsChinese)
{
    std::string topic = getTopic(MessageType::NORMAL, "testMessageBodyContentIsChinese", resource->getBrokerAddr(), resource->getNamesrv(), resource->getCluster());
    std::string group = getGroupId("testMessageBodyContentIsChinese");
    ASSERT_NO_THROW({
        std::shared_ptr<MsgListener> msglistener = std::make_shared<MsgListener>();
        auto pushConsumer = ConsumerFactory::getPushConsumer(topic, group, "*", msglistener);

        auto producer = ProducerFactory::getProducer(group);

        std::string body = "中文字符";
        rocketmq::MQMessage msg(topic, "*", body);
        rocketmq::SendResult sendResult = producer->send(msg);

        ASSERT_EQ(sendResult.getSendStatus(), rocketmq::SendStatus::SEND_OK);

        std::this_thread::sleep_for(std::chrono::seconds(5));

        auto msgs = msglistener->getMessages();

        ASSERT_EQ(msgs.size(), 1);
        ASSERT_EQ(msgs[0].getBody(), body);

        pushConsumer->shutdown();
        producer->shutdown();
    });
}

// Send normal message, setting message body with emoji(😱) character, expect consume success
TEST(MessageBodyContentTest, testMessageBodyContentIsEmoji)
{
    std::string topic = getTopic(MessageType::NORMAL, "testMessageBodyContentIsEmoji", resource->getBrokerAddr(), resource->getNamesrv(), resource->getCluster());
    std::string group = getGroupId("testMessageBodyContentIsEmoji");
    ASSERT_NO_THROW({
        std::shared_ptr<MsgListener> msglistener = std::make_shared<MsgListener>();
        auto pushConsumer = ConsumerFactory::getPushConsumer(topic, group, "*", msglistener);
        std::this_thread::sleep_for(std::chrono::seconds(5));

        auto producer = ProducerFactory::getProducer(group);

        std::string body = "😱";
        rocketmq::MQMessage msg(topic, "*", body);
        rocketmq::SendResult sendResult = producer->send(msg);

        ASSERT_EQ(sendResult.getSendStatus(), rocketmq::SendStatus::SEND_OK);

        std::this_thread::sleep_for(std::chrono::seconds(5));

        auto msgs = msglistener->getMessages();

        ASSERT_EQ(msgs.size(), 1);
        ASSERT_EQ(msgs[0].getBody(), body);

        pushConsumer->shutdown();
        producer->shutdown();
    });
}

File: java/e2e-v4/src/test/java/org/apache/rocketmq/client/message/MessageBodyContentTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.message;

import java.util.concurrent.Callable;

import org.apache.rocketmq.client.rmq.RMQNormalConsumer;
import org.apache.rocketmq.client.rmq.RMQNormalProducer;
import org.apache.rocketmq.enums.TESTSET;
import org.apache.rocketmq.factory.ConsumerFactory;
import org.apache.rocketmq.factory.ProducerFactory;
import org.apache.rocketmq.frame.BaseOperate;
import org.apache.rocketmq.listener.rmq.concurrent.RMQNormalListener;
import org.apache.rocketmq.utils.NameUtils;
import org.apache.rocketmq.utils.VerifyUtils;
import org.junit.jupiter.api.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

/**
 * Test message body
 */
@Tag(TESTSET.CLIENT)
public class MessageBodyContentTest extends BaseOperate {
    private static final Logger log = LoggerFactory.getLogger(MessageBodyContentTest.class);
    private static String topic;
    private RMQNormalProducer producer;
    private RMQNormalConsumer pushConsumer;
    private RMQNormalConsumer pullConsumer;

    @BeforeAll
    public static void setUpAll() {
        topic = getTopic("MessageBodyContentTest");
    }

    @BeforeEach
    public void setUp() {
        
    }

    @AfterEach
    public void tearDown() {
        if (producer != null) {
            producer.shutdown();
        }
        if (pushConsumer != null) {
            pushConsumer.shutdown();
        }
        if (pullConsumer != null) {
            pullConsumer.shutdown();
        }
    }

    @Test
    @DisplayName("Send normal message, setting message body with space character, expect consume success")
    public void testMessageBodyContentIsSpace() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String groupId = getGroupId(methodName);
        String tag = NameUtils.getRandomTagName();

        pushConsumer = ConsumerFactory.getRMQNormalConsumer(namesrvAddr, groupId, rpcHook);
        pushConsumer.subscribeAndStart(topic, tag, new RMQNormalListener());

        producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);

        String body = " ";
        producer.send(topic, tag, body);

        Assertions.assertEquals(1, producer.getEnqueueMessages().getDataSize(), "send message failed");
        await().atMost(60, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return pushConsumer.getListener().getDequeueMessages().getDataSize() == 1;
            }
        });

        VerifyUtils.verifyNormalMessageWithBody(producer.getEnqueueMessages(),
                pushConsumer.getListener().getDequeueMessages(), body);
    }

    @Test
    @DisplayName("Send normal message, setting message body with chinese character, expect consume success")
    public void testMessageBodyContentIsChinese() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String groupId = getGroupId(methodName);
        String tag = NameUtils.getRandomTagName();

        pushConsumer = ConsumerFactory.getRMQNormalConsumer(namesrvAddr, groupId, rpcHook);
        pushConsumer.subscribeAndStart(topic, tag, new RMQNormalListener());

        producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);

        String body = "中文字符";
        producer.send(topic, tag, body);

        Assertions.assertEquals(1, producer.getEnqueueMessages().getDataSize(), "send message failed");

        await().atMost(60, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return pushConsumer.getListener().getDequeueMessages().getDataSize() == 1;
            }
        });

        VerifyUtils.verifyNormalMessageWithBody(producer.getEnqueueMessages(),
                pushConsumer.getListener().getDequeueMessages(), body);
    }

    @Test
    @DisplayName("Send normal message, setting message body with emoji(😱) character, expect consume success ")
    public void testMessageBodyContentIsEmoji() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String groupId = getGroupId(methodName);
        String tag = NameUtils.getRandomTagName();

        pushConsumer = ConsumerFactory.getRMQNormalConsumer(namesrvAddr, groupId, rpcHook);
        pushConsumer.subscribeAndStart(topic, tag, new RMQNormalListener());

        producer = ProducerFactory.getRMQProducer(namesrvAddr, rpcHook);

        String body = "😱";
        producer.send(topic, tag, body);

        Assertions.assertEquals(1, producer.getEnqueueMessages().getDataSize(), "send message failed");

        await().atMost(60, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return pushConsumer.getListener().getDequeueMessages().getDataSize() == 1;
            }
        });

        VerifyUtils.verifyNormalMessageWithBody(producer.getEnqueueMessages(),
                pushConsumer.getListener().getDequeueMessages(), body);
    }
}
